### ISSUE TITLE
TSQL nvarchar(max) incorrectly flagged as L027

### ISSUE THREAD

**Issue by RMHogervorst** [sqlfluff/sqlfluff#2710]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

This nvarchar(max) is flagged L027 (References should be qualified if select has more than one referenced table/view.)
but the max in this case refers to the size of the nvarchar. 

```
CAST(columnname AS NVARCHAR(MAX)) AS columname
```

### Expected Behaviour

I expect the linter to accept this 

### Observed Behaviour

`Unqualified reference 'max' found in select with more than one referenced table/view`  

### How to reproduce

write a query like

```
select
cast(columname as nvarchar(max)) as columname,
anothercolumn
from table1
some join table2
```
You would get this error

### Dialect

tsql

### Version

0.10.1 

### Configuration

sqfluff lint . --dialect tsql --exclude-rules L031

Also a .sqlfluff file present with

```
exclude_rules = L031

[sqlfluff:templater:jinja:context]
variable=thingy
```
but that file is not relevant for the sql file that contains the L027 error

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
I think this is best addressed by changing the `DatatypeSegment` to be a `OneOf` with either `VARCHAR(MAX)` or the current definition:

```py
@tsql_dialect.segment(replace=True)
class DatatypeSegment(BaseSegment):
    """A data type segment.

    Updated for Transact-SQL to allow bracketed data types with bracketed schemas.
    """

    type = "data_type"
    match_grammar = OneOf(
        Sequence("NVARCHAR", Bracketed("MAX")),
        Sequence(
            # Some dialects allow optional qualification of data types with schemas
            Sequence(
                Ref("SingleIdentifierGrammar"),
...
```

Alternatively could just add `MAX` to the Bracketed section of that definition, but then that would allow it to match more (e.g. `VARCHAR(MAX)`).

Either way will need to add `MAX` to list of unreserved keywords for T-SQL. Plus test cases.

**Comment by barrywhart** [comment]
The provided SQL example has a parse error (the word `some`). Use this SQL to test instead:

SQL:
```
select
cast(columname as nvarchar(max)) as columname,
anothercolumn
from table1
join table2
```
